# tarr2fhir API
Convert tarr2fhir Sync4Genes - Versiti and GenDx


### Introduction
This repository contains an API that can translate GenDx xml format to fhir. Input is either a single xml or a collection of all HLA loci xmls for a single Sample/Subject. These input files are generated by Versiti using the Gendx format for Sync4Genes.


### Details
Two API EPs are available 
 1. /tarr/convert2Fhir - this EP accepts a single input Xml generated based on the Tarr.xsd (available in the src/main/resources folder). The xml file will be translated into an Hl7 Fhir bundle in R4 format.
 2. /tarr/convertZip - this EP accepts a zip file containing all the individual HLA locus files for a single sample (or subject)


### Pre-requisites
- Java 8 or above
- Postman or Curl
- If you are a programmer, then an IDE such as one of the following will be useful to modify/update code as needed:
   - IntelliJ 
   - Eclipse
   - Spring Tools Suite (STS) 
- Only item that may need modifying, is the "server.port" in the "application.properties" file. If the default port specified in the application is in use in your local-environment, then the application will need a different "free" port to start up  


### DevOps 
- The code can be checked out from the "master" branch of the "nmdp-bioinformatics/tarr2fhir" page. 
   - If the intent is to add your own code, then kindly fork this repo and create your changes in your forked repo
   - Generate PR when ready and merge to main repository
- The above step will be possible when the commits are merged upstream. Until then, please checkout "jiyer-nmdp/tarr2fhir":
   - master branch will have release ready code before merging upstream
   - develop branch will contain latest development updates   
- After checking out the code, please run the following - 
   - mvn clean
   - mvn install
- Run the application with the following command :
   - java -jar tarr2fhir-0.0.1-SNAPSHOT.jar
- Updates for TARR.xsd schema definition file
   - For privacy reasons, the "TARR.xsd" file is not part of the code repository. 
   - This file needs to be downloaded from - https://www.gendx.com/NGSengine/XSD/20190613/TARR.xsd
   - After downloading the file, please put it in the "src/main/resources/schemas/" folder next to "bindings.xjb".
   - Do this before "mvn install" step listed above


### Testing
- With the application launched locally, the APIs will be available at
  - http://localhost:8099
    - for /tarr/convert2Fhir
      - POST a single gendx format "xml" to this using POSTman or curl
      - Use the following headers: 
        - Content-type: application/xml
    - for /tarr/convertZip
      - POST a zip file containing all HLA locus xml files for a single Sample/Subject using POSTman or curl    
      - Use the following input types:
        - POSTMAN
          - Body: form-data: input file name - "upfile"
                             Choose "File" and upload the zip file
        - CURL
          - curl --location --request POST 'http://localhost:8090/tarr/convertZip' \
            --header 'Access-Control-Request-Headers: *' \
            --form 'upfile=@/Users/jiyer2/Downloads/FHIR NMDP names_03.tarr.zip' 
  - API Responses (for both):
    - 200 OK
      - Json fhir transaction Bundle containing the following resources:
         - Bundle
         - DiagnosticReport
         - Observation (both Allele and Genotype)
         - Provenance
    - if there are errors, standard error messages will be received 
- The API health can be tested by calling one of the two actuator end-points that are enabled as part of spring-boot application setup
  - http://localhost:8099/actuator/health
  - http://localhost:8099/actuator/info    
- The Swagger-Api documentation page can be found at 
  - http://localhost:8099/v2/api-docs 

### TODO
1. The Resources generated only include the following:
   - DiagnosticReport
   - Bundle
   - Observation (both allele and genotype)
   - Provenance
   - #### _Tasks_
        - [ ] Device
        - [ ] Organization
        - [ ] MolecularSequence

2. Update DiagnosticReport and Observation resources with "specimen" reference
   - using an "Extension" that contains - "http://terminology.cibmtr.org/identifier/tarr-sample-name"
   - This is because we are not creating a "Patient" or "Specimen" resource

3. Make the API customizable by providing input "parameters" to specify meta-data options

4. Other items as identified