# tarr2fhir API
Convert tarr2fhir Sync4Genes - Versiti and GenDx


### Introduction
This repository contains an API that can translate an xml file in the GenDx xml format to fhir. These are input xml files generated by Versiti using the Gendx format for Sync4Genes.


### Details
The API currently accepts a single input Xml generated based on the Tarr.xsd (available in the src/main/resources folder). The xml file will be translated into an Hl7 Fhir bundle in R4 format. 


### Pre-requisites
- Java 8 or above
- Postman or Curl
- If you are a programmer, then an IDE such as one of the following will be useful to modify/update code as needed:
   - IntelliJ 
   - Eclipse
   - Spring Tools Suite (STS) 
- Only item that may need modifying, is the "server.port" in the "application.properties" file. If the default port specified in the application is in use in your local-environment, then the application will need a different "free" port to start up  


### DevOps 
- The code can be checked out from the "master" branch of the "nmdp-bioinformatics/tarr2fhir" page. 
   - If the intent is to add your own code, then kindly fork this repo and create your changes in your forked repo
   - Generate PR when ready and merge to main repository
- The above step will be possible when the initial commit is merged. Until then, please checkout "jiyer-nmdp/tarr2fhir":
   - master branch will have release ready code before merging upstream
   - develop branch will contain latest development updates   
- After checking out the code, please run the following - 
   - mvn clean
   - mvn install
- Start the application with the following command :
   - java -jar tarr2fhir-0.0.1-SNAPSHOT.jar


### Testing
- With the application launched locally, the API will be available at
  - http://localhost:8099/tarr/convert2fhir
  - POST a single gendx format "xml" to this using POSTman or curl - the support for "zip" upload will be made available soon
  - Use these headers: 
    - Content-type: application/xml
  - Response:
    - 200 OK
      - Json fhir transaction Bundle containing the following resources:
         - Specimen
         - DiagnosticReport
         - Observation - genotype level
         - Bundle
    - if there are errors, standard error messages will be received 
- The API health can be tested by calling one of the two actuator end-points that are enabled as part of spring-boot application setup
  - http://localhost:8099/actuator/health
  - http://localhost:8099/actuator/info    
- The Swagger-Api documentation page can be found at 
  - http://localhost:8099/v2/api-docs 

### TODO
1. The input right now is a single xml file that contains only a single locus.
    - #### _Tasks_
        - [ ] Ingest a zip file 
        - [ ] Extract all "Gendx format xml" files 
        - [ ] Generate a single Fhir bundle
        - [ ] Possible new API EP for zip file input

2. The Resources generated only include the following:
   - Specimen
   - DiagnosticReport
   - Bundle
   - Observation
   - #### _Tasks_
        - [ ] Observation (both allele level)
        - [ ] Provenance
        - [ ] Device
        - [ ] Organization
        - [ ] MolecularSequence
   
3. Other items as identified
